File: algo-ds-library/python/greedy/activity_selection.py
"""
Activity Selection Problem
---------------------------
Given n activities each with a start and finish time, select the maximum
number of activities that can be performed by a single person, assuming
that a person can only work on one activity at a time.

Greedy Strategy:
    Always pick the activity that finishes earliest — this leaves the most
    room for subsequent activities.

Time Complexity:  O(n log n) — dominated by sorting
Space Complexity: O(n)       — storing selected activities
"""


def activity_selection(activities: list[tuple]) -> list[tuple]:
    """
    Select the maximum number of non-overlapping activities.

    Args:
        activities: List of (start, finish) tuples.

    Returns:
        List of selected (start, finish) tuples in order.

    Example:
        activities = [(1,4), (3,5), (0,6), (5,7), (3,9), (5,9), (6,10), (8,11)]
        → [(1,4), (5,7), (8,11)]
    """
    if not activities:
        return []

    # Sort by finish time — the core greedy decision
    sorted_activities = sorted(activities, key=lambda x: x[1])

    selected = [sorted_activities[0]]
    last_finish = sorted_activities[0][1]

    for start, finish in sorted_activities[1:]:
        # Greedy: take the activity if it starts after the last one finishes
        if start >= last_finish:
            selected.append((start, finish))
            last_finish = finish

    return selected


def activity_selection_with_names(activities: list[dict]) -> list[dict]:
    """
    Activity selection with named activities for readability.

    Args:
        activities: List of {'name': str, 'start': int, 'finish': int} dicts.

    Returns:
        List of selected activity dicts.
    """
    if not activities:
        return []

    sorted_acts = sorted(activities, key=lambda x: x['finish'])
    selected = [sorted_acts[0]]
    last_finish = sorted_acts[0]['finish']

    for act in sorted_acts[1:]:
        if act['start'] >= last_finish:
            selected.append(act)
            last_finish = act['finish']

    return selected

File: java/greedy/ActivitySelection.java
package greedy;

import java.util.*;

/**
 * Activity Selection Problem
 *
 * Greedy strategy: always pick the activity that finishes earliest.
 *
 * Time Complexity:  O(n log n) — sorting dominates
 * Space Complexity: O(n)
 */
public class ActivitySelection {

    public static class Activity {
        public final String name;
        public final int start;
        public final int finish;

        public Activity(int start, int finish) {
            this("", start, finish);
        }

        public Activity(String name, int start, int finish) {
            this.name   = name;
            this.start  = start;
            this.finish = finish;
        }

        @Override
        public String toString() {
            return name.isEmpty()
                ? String.format("(%d,%d)", start, finish)
                : String.format("%s(%d,%d)", name, start, finish);
        }
    }

    /**
     * Select the maximum number of non-overlapping activities.
     * Returns the selected activities in order.
     */
    public static List<Activity> select(List<Activity> activities) {
        if (activities == null || activities.isEmpty()) return Collections.emptyList();

        // Sort by finish time — the greedy choice
        List<Activity> sorted = new ArrayList<>(activities);
        sorted.sort(Comparator.comparingInt(a -> a.finish));

        List<Activity> selected = new ArrayList<>();
        selected.add(sorted.get(0));
        int lastFinish = sorted.get(0).finish;

        for (int i = 1; i < sorted.size(); i++) {
            Activity act = sorted.get(i);
            if (act.start >= lastFinish) {
                selected.add(act);
                lastFinish = act.finish;
            }
        }

        return selected;
    }
}
